doc/source/reference/c-api/array.rst

```diff
@@ -1555,8 +1555,296 @@ For all of these macros *arr* must be an instance of a (subclass of)
+
+   It should set the flags needed for the traversal loop and set *out_loop* to the
+   loop function, which must be a valid ``PyArrayMethod_TraverseLoop``
+   pointer. Currently this is used for zero-filling and clearing arrays storing
+   embedded references.
+
+.. c:type:: int (PyArrayMethod_PromoterFunction)(PyObject *ufunc, \
+                PyArray_DTypeMeta *op_dtypes[], PyArray_DTypeMeta *signature[], \
+                PyArray_DTypeMeta *new_op_dtypes[])
+
+   Type of the promoter function, which must be wrapped into a
+   ``PyCapsule`` with name ``"numpy._ufunc_promoter"``.
+
+   Note that currently the output dtypes are always ``NULL`` unless they are
+   also part of the signature. This is an implementation detail and could
+   change in the future. However, in general promoters should not have a
+   need for output dtypes.
+
+API Functions
+~~~~~~~~~~~~~
+
+These functions are part of the main numpy array API and were added along
+with the rest of the ArrayMethod API.
+
+.. c:function::  int PyUFunc_AddLoopFromSpec( \
+                         PyObject *ufunc, PyArrayMethod_Spec *spec)
+
+   Add loop directly to a ufunc from a given ArrayMethod spec.
+   the main ufunc registration function.  This adds a new implementation/loop
+   to a ufunc.  It replaces `PyUFunc_RegisterLoopForType`.
+
+.. c:function:: int PyUFunc_AddPromoter( \
+                        PyObject *ufunc, PyObject *DType_tuple, PyObject *promoter)
+
+   Register a new promoter for a ufunc. The first argument is the ufunc to
+   register the promoter with. The second argument is a Python tuple containing
+   DTypes or None matching the number of inputs and outputs for the ufuncs. The
+   last argument is a promoter is a function stored in a PyCapsule.  It is
+   passed the operation and requested DType signatures and can mutate it to
+   attempt a new search for a matching loop/promoter.
+
+.. c:function:: int PyUFunc_GiveFloatingpointErrors( \
+                        const char *name, int fpe_errors)
+
+    Signal a floating point error respecting the error signaling setting in
+    the NumPy errstate. Takes the name of the operation to use in the error
+    message and an integer flag that is one of NPY_FPE_DIVIDEBYZERO,
+    NPY_FPE_OVERFLOW, NPY_FPE_UNDERFLOW, NPY_FPE_INVALID to indicate
+    which errors to check for.
+
+    Returns -1 on failure (an error was raised) and 0 on success.
+
+.. c:function:: int PyUFunc_AddWrappingLoop(PyObject *ufunc_obj, \
+            PyArray_DTypeMeta *new_dtypes[], \
+            PyArray_DTypeMeta *wrapped_dtypes[], \
+            PyArrayMethod_TranslateGivenDescriptors *translate_given_descrs, \
+            PyArrayMethod_TranslateLoopDescriptors *translate_loop_descrs)
+
+    Allows creating of a fairly lightweight wrapper around an existing
+    ufunc loop.  The idea is mainly for units, as this is currently
+    slightly limited in that it enforces that you cannot use a loop from
+    another ufunc.
```

Comment:
Would be nice to give a concrete example...

Reply:
(Just to note, I would be fine to not expose this, but it is useful for units, as they should be able to use existing math.)
Not sure if it's easy to make something not very long.  We could point to the unit dtype in the numpy-dtypes repo but that isn't exactly documentation.

Reply:
I'm going to link to an example usage in `metadatadtype`. I also noticed that I forgot to document `PyArrayMethod_TranslateGivenDescriptors` and `PyArrayMethod_TranslateLoopDescriptors` so will document those as well.

Reply:
ended up including a simpler example because linking to the metadatatype code is complicated given that it might change in the future and i'd rather not link to a git hash from today.

Reply:
Super, that example really helps! (and shows my picture of what was happening was wrong...)